nasl/nasl_packet_forgery.c

```diff
@@ -724,6 +732,176 @@ get_tcp_element (lex_ctxt *lexic)
   return retc;
 }
 
+/**
+ * @brief Extract all TCP option from an IP datagram.
+ *
+ * @param[in] options All options present in the TCP segment.
+ * @param[out] tcp_all_options Container for the options to return.
+ */
+static void
+get_tcp_options (char *options, struct tcp_options *tcp_all_options)
+{
+  uint8_t *opt_kind;
+  if (options == NULL)
+    return;
+
+  opt_kind = (uint8_t *) options;
+
+  while (*opt_kind != 0)
+    {
+      switch (*opt_kind)
+        {
+        case TCPOPT_MAXSEG:
+          tcp_all_options->mss.kind = *opt_kind;
+          tcp_all_options->mss.len = *(opt_kind + 1);
+          tcp_all_options->mss.mss = *((uint16_t *) (opt_kind + 2));
+          opt_kind = opt_kind + *(opt_kind + 1);
+          break;
+        case TCPOPT_WINDOW:
+          tcp_all_options->wscale.kind = *opt_kind;
+          tcp_all_options->wscale.len = *(opt_kind + 1);
+          tcp_all_options->wscale.wscale = (uint8_t) * (opt_kind + 2);
+          opt_kind = opt_kind + *(opt_kind + 1);
+          break;
+        case TCPOPT_SACK_PERMITTED:
+          tcp_all_options->sack_perm.kind = *opt_kind;
+          tcp_all_options->sack_perm.len = *(opt_kind + 1);
+          opt_kind = opt_kind + *(opt_kind + 1);
+          break;
+        case TCPOPT_TIMESTAMP:
+          tcp_all_options->tstamp.kind = *opt_kind;
+          tcp_all_options->tstamp.len = *(opt_kind + 1);
+          tcp_all_options->tstamp.tstamp = *((uint32_t *) (opt_kind + 2));
+          tcp_all_options->tstamp.e_tstamp = *((uint32_t *) (opt_kind + 6));
+          opt_kind = opt_kind + *(opt_kind + 1);
+          break;
+        case TCPOPT_EOL:
+        case TCPOPT_NOP:
+          opt_kind++;
+          break;
+        case TCPOPT_SACK: // Not supported
+          opt_kind = opt_kind + *(opt_kind + 1);
+          break;
+        default:
+          g_debug ("%s: Unsupported %u tcp option. "
+                   "Not all options are returned",
+                   __func__, *opt_kind);
+          *opt_kind = 0;
+          break;
+        }
+    }
+}
+
+/**
+ * @brief Get a TCP option from an IP datagram if present.
+ *
+ * @param[in] lexic   Lexical context of NASL interpreter.
+ * @param[in] tcp     The full IP datagram (IP + TCP).
+ * @param[in] option  Option to get.
```

Comment:
maybe add where to get a list of supported options like `see insert_tcp_options()`.